### PR TITLE
Add `PUT /api/groups/{id}` endpoint

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR adds new endpoint that behaves as an "UPSERT" operation for groups. It is part of https://github.com/hypothesis/lms/issues/287

If the `{id}` indicated in the path params corresponds to a group that does not exist, a new group will be created and returned.

If the `{id}` indicated in the path params corresponds to a group that exists and the current user has authority to administer, the group's resource properties will be replaced by the payload of the request (cf. PATCH, which will only update properties that are present in the payload—this will _replace_ the resource, overwriting properties if they are not present in the request). This is to try to make this endpoint behave a little bit more like a PUT verb would intend.

There are a couple of bits of elegance I haven't tackled in this PR for time and complexity sake:

* Some of the logic in the `groups` API view could likely be DRYed out. I want it unrolled for now because we're introducing some new functionality that we've never done before
* The view itself is setting defaults on incomplete PUT payloads in the update branch of this code. We can choose to make this more elegant in an immediate follow-up by putting it somewhere in the schema, or we could decide to wait until another need for this type of thing comes up so that we're not prematurely optimizing.

### CODE REVIEW

File: h/routes.py [hunk 1]
```diff
@@ -112,6 +112,11 @@ def includeme(config):
     config.add_route('api.groups',
                      '/api/groups',
                      factory='h.traversal.GroupRoot')
+    config.add_route('api.group_upsert',
```
Comment (lyzadanger) [inline h/routes.py]: Perhaps unsurprisingly, route ordering matters here. This route will not match correctly if it is defined after the `api.group` route (as that route would match, too, but has no handler for `PUT`).

File: h/views/api/groups.py [hunk 2]
```diff
@@ -95,6 +95,61 @@ def update(group, request):
...
+    update_properties = {
+        'name': appstruct['name'],
+        'description': appstruct.get('description', ''),
+        'groupid': appstruct.get('groupid', None),
+    }
+
+    group = group_update_service.update(group, **update_properties)
+
+    # Note that this view takes a ``GroupUpsertContext`` but uses a ``GroupContext`` here
+    return GroupJSONPresenter(GroupContext(group, request)).asdict(expand=['organization'])
```
Comment (lyzadanger) [inline h/views/api/groups.py]: One bothersome thing about this endpoint, which we may want to address in the short term (this is a brokenness across our whole API), is that both operations—create or update—will return an HTTP 200. The case should be that our `create` returns a 201 and our `update` returns a 200.

File: h/routes.py (hunk 1 above)
Comment (lyzadanger) [reply h/routes.py]: To protect ourselves here, we _could_ for now put an explicit `request_method='PATCH'` on the `api.group` route).

Comment (seanh) [reply h/routes.py]: I think the explicit request method would be more Pythonic than implicitly relying on ordering

File: h/views/api/groups.py [hunk 3]
```diff
@@ -95,6 +95,61 @@ def update(group, request):
...
+    Create or update a group from a PUT payload.
+
+    If no group model is present in the passed ``context`` (on ``context.group``),
+    treat this as a create action and delegate to ``create``.
+
+    Otherwise, replace the existing group's resource properties entirely and update
+    the object.
+
+    :arg context:
+    :type context: h.traversal.contexts.GroupUpsertContext
```
Comment (seanh) [inline h/views/api/groups.py]: This should be `h.traversal.GroupUpsertContext` (the fact that `h.traversal` is broken down into `roots` and `contexts` modules is an internal implementation detail of that package)

File: tests/h/views/api/groups_test.py [hunk 4]
```diff
@@ -311,6 +311,143 @@ def group(self, factories):
...
+                         'CreateGroupAPISchema',
+                         'group_service',
+                         'group_update_service',
+                         'GroupContext',
+                         'GroupJSONPresenter',)
+class TestUpsertGroup(object):
+
+    def test_it_proxies_to_create_if_group_empty(self, pyramid_request, create):
+        empty_group_context = mock.Mock()
+        empty_group_context.group = None
```
Comment (seanh) [inline tests/h/views/api/groups_test.py]: This should be `create_autospec()` (we should basically never construct bare `Mock()`s or `MagicMock()`s)

File: tests/h/views/api/groups_test.py [hunk 5]
```diff
@@ -311,6 +311,143 @@ def group(self, factories):
...
+    @pytest.fixture
+    def group(self, factories, group_user):
+        return factories.Group(creator=group_user)
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request, factories):
+        # Add a nominal json_body so that _json_payload() parsing of
+        # it doesn't raise
+        pyramid_request.json_body = {}
+        # pyramid_request.user = factories.User()
```
Comment (seanh) [inline tests/h/views/api/groups_test.py]: Looks like a commented-out line got left in here

Comment (seanh) [reply tests/h/views/api/groups_test.py]: (And the `factories` arg can be removed)

File: tests/h/views/api/groups_test.py [hunk 6]
```diff
@@ -311,6 +311,143 @@ def group(self, factories):
...
+        group_context = mock.Mock()
+        group = factories.Group(authority_provided_id='something_else', authority='example.com')
+        group_context.group = group
+        group_service.fetch.return_value = group
+
+        views.upsert(group_context, pyramid_request)
+
+    def test_it_proxies_to_update_service_with_injected_defaults(self,
+                                                                 pyramid_request,
+                                                                 group_service,
```
Comment (seanh) [inline tests/h/views/api/groups_test.py]: `group_service` is unused

File: tests/h/views/api/groups_test.py [hunk 7]
```diff
@@ -311,6 +311,143 @@ def group(self, factories):
...
+
+        CreateGroupAPISchema.return_value.validate.assert_called_once_with({
+            'name': 'Rename Group'
+        })
+
+    def test_it_raises_ConflictError_on_duplicate(self,
+                                                  pyramid_request,
+                                                  group_service,
+                                                  factories,
+                                                  CreateGroupAPISchema):
```
Comment (seanh) [inline tests/h/views/api/groups_test.py]: `CreateGroupAPISchema` is unused

File: tests/h/views/api/groups_test.py [hunk 8]
```diff
@@ -311,6 +311,143 @@ def group(self, factories):
...
+        group_context = mock.Mock()
+        group_context.group = group
+
+        views.upsert(group_context, pyramid_request)
+
+        assert create.call_count == 0
+
+    def test_it_validates_against_group_update_schema_if_group_extant(self,
+                                                                      pyramid_request,
+                                                                      create,
```
Comment (seanh) [inline tests/h/views/api/groups_test.py]: `create` is unused

**Pull request merged**